### DISCUSSION TITLE
@computed_field not showing in OpenAPI Json

### DISCUSSION THREAD

**Discussion by Galarzaa90** [fastapi/fastapi#9856] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import pydantic
import uvicorn

class StudentGrades(pydantic.BaseModel):
    english: int
    history: int
    geography: int

    @pydantic.computed_field()
    @property
    def grade(self) -> float:
        return (self.english + self.history + self.geography)/ 3
    

app = FastAPI()

@app.get("/")
def sample() -> StudentGrades:
    return StudentGrades(english=100, history=100, geography=70)
```

### Description

When using model with a computed field, if I check the content of the response:

```jsonc
{
  "english": 100,
  "history": 100,
  "geography": 70,
  "grade": 90.0   // computed field
}
```

The computed field is in the response, as expected.

However, the computed field is not being added to the json schema:
```json
{
  "openapi": "3.1.0",
  "info": {
    "title": "FastAPI",
    "version": "0.1.0"
  },
  "paths": {
    "/": {
      "get": {
        "summary": "Sample",
        "operationId": "sample__get",
        "responses": {
          "200": {
            "description": "Successful Response",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/StudentGrades"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "StudentGrades": {
        "properties": {
          "english": {
            "type": "integer",
            "title": "English"
          },
          "history": {
            "type": "integer",
            "title": "History"
          },
          "geography": {
            "type": "integer",
            "title": "Geography"
          }
        },
        "type": "object",
        "required": [
          "english",
          "history",
          "geography"
        ],
        "title": "StudentGrades"
      }
    }
  }
}
```

And therefore, not showing in Swagger UI:
![image](https://github.com/tiangolo/fastapi/assets/12865379/180eca85-74d8-4e0e-a41d-10c46ae67eb2)

### Operating System

Windows, macOS

### FastAPI Version

0.100.0

### Python Version

Python 3.10.11

### Additional Context

I initially thought this was an issue on [Pydantic](https://github.com/pydantic/pydantic), and i even opend an [issue](https://github.com/pydantic/pydantic/issues/6564)

But as explained there, the computed field appears when using `model_json_schema(mode='serialization')`, and not without it.

**Comment by ac130kz** [comment]
`cached_property` shows up in the example, but not in the schema for me. Why not make this an issue?

**Comment by Kludex** [comment, marked as answer]
@tiangolo is already aware of this. This will shortly be supported.

**Reply by Kludex** [reply]
This is not a bug, it's a feature request, since in Pydantic V1 there was no `@computed_field`. Which, once again, it will be available soon in FastAPI.

**Reply by hueindahaus** [reply]
What is a rough eta on this? Days? Weeks? Months?

**Reply by Kludex** [reply]
I don't know.

**Reply by tiangolo** [reply]
This is now supported in FastAPI 0.101.0 🎉

**Reply by carlosdorneles-mb** [reply]
In FastAPI 0.111.0, fields with the computed_field decorator are not appearing in the documentation.

```python
    @computed_field(description="Operation type group.", examples=OperationGroupTypeEnum.get_values())
    @property
    def type_group(self) -> str:
        return self.type.group
```

**Reply by asparagusbeef** [reply]
> In FastAPI 0.111.0, fields with the computed_field decorator are not appearing in the documentation.
> 
> ```python
>     @computed_field(description="Operation type group.", examples=OperationGroupTypeEnum.get_values())
>     @property
>     def type_group(self) -> str:
>         return self.type.group
> ```

I am having a reverse issue. the computed_field appears in the documentation but not in the actual response.

**Reply by filias** [reply]
> In FastAPI 0.111.0, fields with the computed_field decorator are not appearing in the documentation.
> 
> ```python
>     @computed_field(description="Operation type group.", examples=OperationGroupTypeEnum.get_values())
>     @property
>     def type_group(self) -> str:
>         return self.type.group
> ```

This is still valid for me. Will there be any fix or do we need to change anything in our code?

**Reply by xiaoling** [reply]
> In FastAPI 0.111.0, fields with the computed_field decorator are not appearing in the documentation.
> 
> ```python
>     @computed_field(description="Operation type group.", examples=OperationGroupTypeEnum.get_values())
>     @property
>     def type_group(self) -> str:
>         return self.type.group
> ```

and not in the response of `/api/v1/openapi.json`
cc @tiangolo @Kludex  🙏

**Reply by PHvL** [reply]
> In FastAPI 0.111.0, fields with the computed_field decorator are not appearing in the documentation.
> 
> ```python
>     @computed_field(description="Operation type group.", examples=OperationGroupTypeEnum.get_values())
>     @property
>     def type_group(self) -> str:
>         return self.type.group
> ```

Are you using `separate_input_output_schemas=False`? This seems to use validation instead of serialization schema.

**Reply by undera** [reply]
Regardless of `separate_input_output_schemas`, it reproduces for me on 0.115.6

**Reply by danielborges93** [reply]
Here I worked around this issue in version 0.115.11 as follows:
```python
class StudentGrades(pydantic.BaseModel):
    english: int
    history: int
    geography: int
    grade: float = Field(
        default_factory=lambda data: (data['english'] + data['history'] + data['geography']) / 3,
    )
```

**Reply by yamork779** [reply]
@danielborges93 
it reproduces on 0.115.12 when separate_input_output_schemas=false

example:

``` 
app = FastAPI(
    separate_input_output_schemas=False,
    root_path="/api"
)

class Example(BaseModel):
    _value: Any = PrivateAttr()
    @computed_field(alias="value", title="value", repr=True, description="test")
    @property
    def value(self) -> Any:
        return self._value
    @value.setter
    def value(self, value: Any):
        self._value = value


@app.get("/test", summary="test")
def test() -> Example:
    pass
```

**Comment by rossmacarthur** [comment]
In the API route `__init__` the following code is called for 

```python
            self.response_field = create_response_field(
                name=response_name,
                type_=self.response_model,
                # TODO: This should actually set mode='serialization', just, that changes the schemas
                # mode="serialization",
                mode="validation",
            )
```

Is it possible `mode` could be configurable? Or is the plan to make it the default

**Reply by tiangolo** [reply]
Yep, that was the change in the internal code ...plus some other thousand lines of test updates.

**Reply by slafs** [reply]
`mode` _is_ configurable, but on a model level - see https://github.com/fastapi/fastapi/discussions/10697#discussioncomment-10049975 for a workaround

**Comment by TiansuYu** [comment]
It has been a year, what is the current status?

**Reply by TiansuYu** [reply]
Tested against `fastapi=0.111`. Seems to be working with the response payload (and from the code of https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L472, it looks like has been patched). But the computed field still does not show up in the REST API doc.

**Reply by DavideGalilei** [reply]
Current PRs addressing this issue:
- (tests failing): https://github.com/fastapi/fastapi/pull/12119
- (tests passing, not merged yet): https://github.com/fastapi/fastapi/pull/11517

**Comment by DavideGalilei** [comment]
A temporary workaround for this issue: https://github.com/fastapi/fastapi/discussions/10697#discussioncomment-10049975

**Reply by espdev** [reply]
> A temporary workaround for this issue: [#10697 (comment)](https://github.com/fastapi/fastapi/discussions/10697#discussioncomment-10049975)

With `model_config = ConfigDict(json_schema_mode_override='serialization')` I get resolver errors for each field in models in OpenAPI docs when I unfold an API method:
![2025-02-20_11-56](https://github.com/user-attachments/assets/7fbaa61a-5bd0-456e-a960-fcaf83356566)

So, it seems it is not a workaround.

**Reply by mattzque** [reply]
just ran into the same issue, the workaround with json_schema_mode_override=serialization didn't work for me either. What did work was to manually add the property:

```python
def json_schema_add_computed(schema):
    schema['properties']['my_computed_field'] = ...
...
model_config = ConfigDict(json_schema_extra=json_schema_add_computed)
```

I wonder if the schema errors are caused by setting `separate_input_output_schemas=False` on FastAPI.

**Reply by jb3rndt** [reply]
In my case it was a matter of the order of the annotations. The `computed_field` annotation has to go first. So this is wrong:
```python
@property
@computed_field
def some_field(self):
    return None
```

This is correct:
```python
@computed_field
@property
def some_field(self):
    return None
```

**Reply by jackpmastrangelo** [reply]
@mattzque I agree, I also believe this is related to `separate_input_output_schemas=False`
